boot/LoadReactOS.c

```diff
@@ -0,0 +1,410 @@
+		printk("No multiboot header found\n");
+		return NULL;
+	}
+
+	return mbHeader;
+}
+
+CONFIGENTRY *DetectReactOSFATX(FATXPartition *partition) {
+	FATXFILEINFO fileinfo;
+	CONFIGENTRY *config = NULL;
+
+	if (LoadFATXFile(partition, "/freeldr.sys", &fileinfo)) {
+		// Root of E has a freeldr.sys in
+		config = (CONFIGENTRY *)malloc(sizeof(CONFIGENTRY));
+		memset(config, 0x00, sizeof(CONFIGENTRY));
+		config->bootSystem = SYS_REACTOS;
+		strcpy(config->title, "ReactOS");
+		strcpy(config->szPath, "/freeldr.sys");
+
+		free(fileinfo.buffer);
+	}
+
+	return config;
+}
+
+int LoadReactOSFATX(FATXPartition *partition, CONFIGENTRY *config) {
+	static FATXFILEINFO fileinfo;
+	OPTMULTIBOOT *multiboot = &config->opt.Multiboot;
+
+	memset(&fileinfo, 0x00, sizeof(fileinfo));
+
+	VIDEO_ATTR = 0xffd8d8d8;
+	printk("  Loading %s from FATX", config->szPath);
+	if (!LoadFATXFilefixed(partition, config->szPath, &fileinfo, FREELDR_LOAD_AREA)) {
+		printk("Error loading %s\n", config->szPath);
+		wait_ms(2000);
+		return false;
+	} else {
+		printk(" - %d bytes\n", fileinfo.fileRead);
+
+		if (CheckMultibootHeader(fileinfo.buffer) == NULL) {
+			wait_ms(2000);
+			return false;
+		}
+		multiboot->pBuffer = fileinfo.buffer;
+		multiboot->uBufferSize = fileinfo.fileRead;
+		/* Multiboot partition indices starting at 0, FATX (E:) is first one */
+		multiboot->uBootDevice = GetMultibootBootDevice(0x80 + partition->nDriveIndex, 0, 0xFF, 0xFF);
+	}
+
+	return true;
+}
+
+CONFIGENTRY *DetectReactOSCD(int cdromId) {
+	long dwSize;
+	CONFIGENTRY *config;
+
+	dwSize = BootIso9660GetFile(cdromId, "/loader/setupldr.sys", FREELDR_LOAD_AREA, FREELDR_MAX_SIZE);
+
+	// Failed to load freeloader
+	if (dwSize <= 0)
+		return NULL;
+
+	// Freeloader found
+	config = (CONFIGENTRY *)malloc(sizeof(CONFIGENTRY));
+	memset(config, 0x00, sizeof(CONFIGENTRY));
+	config->bootSystem = SYS_REACTOS;
+	strcpy(config->title, "ReactOS");
+	strcpy(config->szPath, "/loader/setupldr.sys");
+
+	return config;
+}
+
+int LoadReactOSCD(CONFIGENTRY *config) {
+	long dwSize;
+	OPTMULTIBOOT *multiboot = &config->opt.Multiboot;
+
+	memset(FREELDR_LOAD_AREA, 0, FREELDR_MAX_SIZE);
+
+	VIDEO_ATTR = 0xffd8d8d8;
+	printk("  Loading %s from CD", config->szPath);
+	VIDEO_ATTR = 0xffa8a8a8;
+	dwSize = BootIso9660GetFile(config->drive, config->szPath, FREELDR_LOAD_AREA, FREELDR_MAX_SIZE);
+
+	if (dwSize < 0) {
+		printk("Not Found, error %d\nHalting\n", dwSize); 
+		wait_ms(2000);
+		return false;
+	} else {
+		printk(" - %d bytes\n", dwSize);
+
+		if (CheckMultibootHeader(FREELDR_LOAD_AREA) == NULL) {
+			wait_ms(2000);
+			return false;
+		}
+		multiboot->pBuffer = FREELDR_LOAD_AREA;
+		multiboot->uBufferSize = dwSize;
+		/* Multiboot partition indices starting at 0, FreeLoader will increment 0xFE to 0xFF
+		 * and later it will interpret partition number 0xFF as CD drive */
+		multiboot->uBootDevice = GetMultibootBootDevice(0x80 + config->drive, 0xFE, 0xFF, 0xFF);
```

Comment:
See comment.

Reply:
@disean fixed, thanks! Is the partition number ok, or it should be equal to 0?

Reply:
That's correct now.